### DISCUSSION TITLE
Private annotations on public records

### DISCUSSION THREAD

**Discussion by intrnl** [bluesky-social/atproto#1672] [Bluesky Lexicons]
**Is your feature request related to a problem? Please describe.**

Users needs a way to add private notes regarding blocks

**Describe the solution you'd like**

It would be nice if we could have a way to privately annotate an otherwise public record, mainly for blocks, but can be extended for other uses as well (like/follow/repost reasoning, drafts for published posts, etc)

We introduce a new field on `record` type named `annotation`, this is a schema in which private annotations can be validated.

```json
{
  "lexicon": 1,
  "id": "app.bsky.graph.block",
  "defs": {
    "main": {
      "type": "record",
      "key": "tid",
      "record": {},
      "annotation": {
        "type": "object",
        "required": [],
        "properties": {
          "notes": { "type": "string" }
        }
      }
    }
  }
}
```

Annotation can be done by including them alongside the actual record value in `createRecord` method

```ts
await agent.com.atproto.repo.createRecord({
  repo,
  collection: "app.bsky.graph.block",
  record: {
    createdAt: new Date().toISOString(),
    subject: "did:web:example.com"
  },
  annotation: {
    notes: "Insert notes here explaining the block"
  }
});
```

**Describe alternatives you've considered**

TODO: Bossett over on the API Touchers Discord server suggested the idea of a notes repository that's only visible to the user, I've considered this, but feel that having to manage two separate records, one a "public record" and one a "private record annotating the public record" seems finicky, but we could try, I'll put up some examples here perhaps

**Additional context**

None.

**Comment by mackuba** [comment]
It would also be nice to be able to add private annotations on other profiles, to keep a private note about someone (who isn't necessarily blocked). Same way as Mastodon does it:

<img width="549" src="https://github.com/bluesky-social/atproto/assets/28465/75e12d93-d302-4e42-b2a0-ee1fefa610f2">

**Comment by devinivy** [comment]
This is a very cool idea, converting to a discussion to continue fleshing it out.

**Comment by DavidBuchanan314** [comment]
I like the idea of a "private" repo alongside the public one - this would address my open question of how best to export all non-public data during account migrations https://github.com/bluesky-social/atproto/discussions/1142#discussioncomment-7112939

**Reply by pfrazee** [reply]
Private data won't solidify until post-federation, but this is what I'm advocating for too

**Comment by ghost** [comment]
Private notes are useful for mutes, numbers, associated links, reminders, blocks,  I use the notes feature on discord for almost every contact. Saves Time too

**Comment by pfrazee** [comment]
> TODO: Bossett over on the API Touchers Discord server suggested the idea of a notes repository that's only visible to the user, I've considered this, but feel that having to manage two separate records, one a "public record" and one a "private record annotating the public record" seems finicky, but we could try, I'll put up some examples here perhaps

Bossett is correct, it's not possible to co-mingle private data into the public repo. By nature, the public repo has to be 100% public; records or parts of records can't be private.

The private data story is something that will be addressed post-federation. We had extensive discussions about it last summer and have a good set of ideas for how we'll do it, but it has to wait until the public network is fully operating.

**Comment by est** [comment]
any updates on this?

Keeping a why/when remark on the follower/following relation data would be helpful

https://github.com/bluesky-social/atproto/issues/2981
